### PR TITLE
[DCOS_OSS-444] Introduce toggle for TLS 1.0 support in Admin Router

### PR DESCRIPTION
**Pull request by adamtheturtle**
## High Level Description

* Change the default TLS configuration.
  By default before, TLS 1.0 was enabled.
  By default with this change, TLS 1.0 is disabled.

* Allow toggling TLS 1.0.
  This is with a configuration variable.

## Related Issues

  - [DCOS_OSS-1819](https://jira.dcos.io/browse/DCOS_OSS-1819) Document that Admin Router does not support TLS 1.0 by default anymore.
  - [DCOS_OSS-444](https://jira.dcos.io/browse/DCOS_OSS-444) Admin Router: disable TLS 1.0 support by default.
- [DCOS-13437](https://jira.mesosphere.com/browse/DCOS-13437) Add DC/OS config option to enable Admin Router serving TLS 1.0 (after disabling by default)

## Checklist for all PR's

  - [X] Included a test which will fail if code is reverted but test is not.
  - [X] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [X] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Lower Level Changes

* Move `gen` tests into a `tests` directory.
  This separates them slightly from implementation files but for this change I chose to do it here in order to have a new file for common `gen` testing utilities.

* Move TLS configuration to a generated file.
  This was in the static `packages/adminrouter/extra/src/common/server.conf` file.
  This is now in the generated `adminrouter-tls.conf` file.
  Only the SSL protocols line needs to be in a generated file.
  However, it seemed sensible to keep all TLS configuration together.

* Add a "realistic" saple configuration for the new config file to the Admin Router test Docker container.

## Further considered work

* Rather than instructing on manual tests, add these to CI.
  That relies on test infrastructure which is not yet built.
  That test infrastructure would associate configurations with tests and create clusters.

* Rather than a textual description of commands to run, create a small script.
  I think that the effort is better spent on the aforementioned test infrastructure.
  The manual steps are very simple.

* Add a test that the Admin Router Docker container has a "realistic" configuration.
  That is, add a test which would fail if the new `packages/adminrouter/docker/adminrouter-tls.conf` sample did not exist.
  I am not sure of the trade-offs involved with doing this.

* The comments referred to in the code are in the private JIRA.
  Perhaps move these and the references to the public JIRA.

I am happy to split this up as a reviewer wishes.

## Manual testing

See comments on https://jira.mesosphere.com/browse/DCOS-13437 for a detailed description of the manual testing and the motivations behind it.

I ran the following commands on the `master` branch and this branch to demonstrate the difference between the two:

 ```
 # I ran the following with tls1, tls1_1 and tls1_2 on master nodes
 dcos-shell openssl s_client -connect localhost:443 -tls1
 
 # I ran the following with tls1, tls1_1 and tls1_2 on agent nodes
 dcos-shell openssl s_client -connect localhost:61002 -tls1
 ```
 
 I considered "Works" to be that the output included "Connected" and a certificate.
 I considered "Not supported" to be that the output included "Connected" and no certificate.

 I got the following results:
 
- Master Branch:
	- Master node:
		- TLS 1.0 - Works
		- TLS 1.1 - Works
		- TLS 1.2 - Works
	- Agent node:
		- TLS 1.0 - Connection Refused
		- TLS 1.1 - Connection Refused
		- TLS 1.2 - Connection Refused
- New branch TLS 1.0 enabled:
	- Master node:
		- TLS 1.0 - Works
		- TLS 1.1 - Works
		- TLS 1.2 - Works
	- Agent node:
		- TLS 1.0 - Connection Refused
		- TLS 1.1 - Connection Refused
		- TLS 1.2 - Connection Refused
- New branch TLS 1.0 disabled:
	- Master node:
		- TLS 1.0 - Not supported
		- TLS 1.1 - Works
		- TLS 1.2 - Works
	- Agent node:
		- TLS 1.0 - Connection Refused
		- TLS 1.1 - Connection Refused
		- TLS 1.2 - Connection Refused

## Other information

Evidence that this does not break DC/OS Enterprise is at https://github.com/mesosphere/dcos-enterprise/pull/803/files.

### CODE REVIEW

File: gen/tests/utils.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+"""
+Utilities for tests for ``gen``.
+"""
+
+import copy
```
Comment (adamtheturtle) [inline gen/tests/utils.py]: Please note in review that these utilities are not new. They are instead moved from `test_validation.py`.

**Comment by adamtheturtle** [comment]
@mesosphere-mergebot label Ready For Review

**Comment by adamtheturtle** [comment]
cc @vespian @mhrabovcin @jgehrcke for review

**Comment by vespian** [comment]
> [DCOS_OSS-444] Disable admin router config dcos 13437

This is a bit imprecise, could you please clarify it? E.G "Add an option to disable TLS1.0 protocol support for AR"

```
OSS:
Master Branch:
Master node:
TLS 1.0 - Works
TLS 1.1 - Works
TLS 1.2 - Works
Agent node:
TLS 1.0 - Works
TLS 1.1 - Works
TLS 1.2 - Works
```

I am curious how you got this results  - Open AR does not support TLS :|

https://github.com/dcos/dcos/blob/b9e80193a419cb016725a54b4e4d0efdaf5803f7/gen/dcos-config.yaml#L905-L910

---

Reviewed 8 of 19 files at r2, 1 of 3 files at r3, 3 of 7 files at r4, 3 of 5 files at r5.
Review status: 12 of 16 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[gen/dcos-config.yaml, line 905 at r5](https://reviewable.io:443/reviews/dcos/dcos/1448#-Ki0Yi1fq2egFrrvMXOt:-Ki0Yi1gVfwiVPtXaLlo:biu6bq8) ([raw file](https://github.com/dcos/dcos/blob/6efa6acfafeac52d8f3e1401651332322bd42664/gen/dcos-config.yaml#L905)):*
> ```YAML
>       ssl_certificate common/snakeoil.crt;
>       ssl_certificate_key common/snakeoil.key;
>   - path: /etc/adminrouter-tls.conf
> ```

Thanks for separating all TLS stuff into a separate file! 

---

*[gen/tests/test_adminrouter_tls_conf.py, line 78 at r5](https://reviewable.io:443/reviews/dcos/dcos/1448#-Ki0_UIVRvuxEWUWRGJr:-Ki0_UIVRvuxEWUWRGJs:bkh18rj) ([raw file](https://github.com/dcos/dcos/blob/6efa6acfafeac52d8f3e1401651332322bd42664/gen/tests/test_adminrouter_tls_conf.py#L78)):*
> ```Python
>         [ssl_protocols_line] = [
>             line for line in config['content'].split('\n') if
>             line.startswith('ssl_protocols ')
> ```

line.lstrip().startswith(...) ?

---

*[gen/tests/test_adminrouter_tls_conf.py, line 96 at r5](https://reviewable.io:443/reviews/dcos/dcos/1448#-Ki1EV5qS118elo1ijI3:-Ki1EV5ril6rOjyqXlya:b-p909ez) ([raw file](https://github.com/dcos/dcos/blob/6efa6acfafeac52d8f3e1401651332322bd42664/gen/tests/test_adminrouter_tls_conf.py#L96)):*
> ```Python
>     @pytest.mark.parametrize(
>         'new_arguments', [{}, {'adminrouter_tls_1_0_enabled': 'false'}]
>     )
> ```

Why parametrize it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dcos/dcos/1448#-:-Ki0Q-u2Wxxv_EHzEGbr:bzbwyzv)*

**Comment by vespian** [comment]
Reviewed 5 of 19 files at r2, 1 of 3 files at r3, 1 of 5 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dcos/dcos/1448)*

**Comment by adamtheturtle** [comment]
```
I am curious how you got this results - Open AR does not support TLS :|

https://github.com/dcos/dcos/blob/b9e80193a419cb016725a54b4e4d0efdaf5803f7/gen/dcos-config.yaml#L905-L910
```

Eurgh sorry that is an error made when changing from "Works" = "As expected" to "Works" = "TLS connection works". I am sure that this was all as expected and have changed the PR description.

File: gen/tests/test_adminrouter_tls_conf.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        The config variable `tls_1_0_enabled` must be 'true' or 'false'.
+        """
+        validate_error(
+            new_arguments={'adminrouter_tls_1_0_enabled': 'foo'},
+            key='adminrouter_tls_1_0_enabled',
+            message=true_false_msg,
+        )
+
+    @pytest.mark.parametrize(
+        'new_arguments', [{}, {'adminrouter_tls_1_0_enabled': 'false'}]
```
Comment (adamtheturtle) [inline gen/tests/test_adminrouter_tls_conf.py]: > @vespian : Why parametrize it?

My aim was to make it clear that `{}` and `{'adminrouter_tls_1_0_enabled': 'false'}` are the same. That is, `{'adminrouter_tls_1_0_enabled': 'false'}` is the default.

Comment (vespian) [inline gen/tests/test_adminrouter_tls_conf.py]: ah... kk. Just please add a comment to clarify this.

Comment (adamtheturtle) [inline gen/tests/test_adminrouter_tls_conf.py]: Ok, I thought that the doctoring was that comment but I will expand it.

File: gen/tests/test_adminrouter_tls_conf.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            A ``list`` of supported SSL protocols.
+        """
+        arguments = make_arguments(new_arguments=new_config_arguments)
+        generated = gen.generate(arguments=arguments)
+        package = generated.templates['dcos-config.yaml']['package']
+        config_path = '/etc/adminrouter-tls.conf'
+        [config] = [item for item in package if item['path'] == config_path]
+        [ssl_protocols_line] = [
+            line for line in config['content'].split('\n') if
+            line.startswith('ssl_protocols ')
```
Comment (adamtheturtle) [inline gen/tests/test_adminrouter_tls_conf.py]: > @vespian : line.lstrip().startswith(...) ?

Why? Is is possible that there are two lines with `ssl_protocols` but one starts with whitespace?

I can make this change but I think that that is obscure and partially covered by `test_default` Would you still like the change?

**Comment by adamtheturtle** [comment]
> [DCOS_OSS-444] Disable admin router config dcos 13437
This is a bit imprecise, could you please clarify it? E.G "Add an option to disable TLS1.0 protocol support for AR"

Are you OK to consider this to be the OSS version of https://jira.mesosphere.com/browse/DCOS-13437?

File: gen/tests/test_adminrouter_tls_conf.py (hunk 2 above)
Comment (vespian) [inline gen/tests/test_adminrouter_tls_conf.py]: Look at it from a perspective of a simple soldier like me - I did not immediately understand it and fortunately, you were here to help me. Somebody else, some time in the future may have the same problem like me and you will not be there to help him :) The docstring at least for me is not immediately obvious neither this decorator.

Comment (adamtheturtle) [inline gen/tests/test_adminrouter_tls_conf.py]: Yep that's why I put "I will expand it." 😄 I was just saying "I tried"

**Comment by adamtheturtle** [comment]
Thank you for the review @vespian . I have addressed all of the comments.

**Comment by adamtheturtle** [comment]
Please note that after this said that all tests were passing a bunch of new, failing tests appeared. I'm not sure if they are relevant and would like reviewer feedback.

File: gen/tests/test_adminrouter_tls_conf.py (hunk 2 above)
Comment (vespian) [inline gen/tests/test_adminrouter_tls_conf.py]: :+1:

File: gen/tests/test_adminrouter_tls_conf.py (hunk 3 above)
Comment (vespian) [inline gen/tests/test_adminrouter_tls_conf.py]: Yes please.

I am a bit paranoid and prefer to not leave any loopholes - people can get very creative at making bugs ;) I do not think we could end up with multiple `ssl_protocols` lines right now/with current code. But we do not know what future maintainers are going to come up with and adding `.lstrip()` here is a very cheap trap... I mean extra test we can prepare for them :D

File: gen/tests/utils.py (hunk 1 above)
Comment (vespian) [inline gen/tests/utils.py]: Thanks for saying it!

File: gen/tests/test_adminrouter_tls_conf.py (hunk 3 above)
Comment (adamtheturtle) [inline gen/tests/test_adminrouter_tls_conf.py]: FWIW I did this yesterday the assumption that you'd say yes 😄

File: gen/tests/test_adminrouter_tls_conf.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        """
+        config_path = '/etc/adminrouter-tls.conf'
+        arguments = make_arguments(new_arguments={})
+        generated = gen.generate(arguments=arguments)
+        package = generated.templates['dcos-config.yaml']['package']
+        [config] = [item for item in package if item['path'] == config_path]
+
+        expected_configuration = dedent(
+            """\
+            # Ref: https://github.com/cloudflare/sslconfig/blob/master/conf
```
Comment (vespian) [inline gen/tests/test_adminrouter_tls_conf.py]: NIT: do we want to match comments here as well?

Comment (adamtheturtle) [inline gen/tests/test_adminrouter_tls_conf.py]: Two things:

1. As per the docstring this is mainly a sanity check. Having extra logic here to remove comments from the check is probably more hassle than having to change this test whenever there is a comment change.

2. During development I accidentally put the manual test comment inside a conditional and this caught it.

tl;dr: yes, IMO

File: gen/tests/test_adminrouter_tls_conf.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    `openssl` commands.
+
+    See comments on https://jira.mesosphere.com/browse/DCOS-13437 for more
+    details.
+    """
+
+    def supported_ssl_protocols(self, new_config_arguments) -> List[str]:
+        """
+        This finds a line which looks like the following:
+            ssl protocols TLSv1, TLSv1.1;
```
Comment (jgehrcke) [inline gen/tests/test_adminrouter_tls_conf.py]: Missing an underscore :)

**Comment by jgehrcke** [comment]
@mesosphere-mergebot label ship it

**Comment by orsenthil** [comment]
@mesosphere-mergebot label ship it

**Comment by orsenthil** [comment]
@jgehrcke  - need to be an owner for the "ship-it" sauce.  (Ref: https://github.com/dcos/dcos/pull/1448#issuecomment-293246860) I'll update the file to include your login.
